AddFeature Modification

s9latimm opened the pull request.

s9latimm committed:
Implement AddFeatureToModel

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp

s9latimm committed:
Use unitests fixtures

M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/NumericFeature.cpp
M	unittests/Feature/OrderedFeatureVector.cpp
M	unittests/Feature/Relationship.cpp

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -154,28 +153,29 @@ class FeatureModelBuilder : private FeatureModel {
-
-  FeatureModelBuilder *addParent(const std::string &FeatureName,
-                                 const std::string &ParentName) {
+  FeatureModelBuilder *addEdge(const std::string &ParentName,
```
Why did you rename this? `addParent` seemed fine to me

s9latimm replied:
I wanted to clarify, that calling this implies, that the children relation between these nodes will be updated too (while building the tree) -- in contrast to the `FeatureModelModification` where we implemented `setParent` so it literally changes only that.

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -135,14 +135,13 @@ class FeatureModelModification {
+  static void setParent(Feature &F, Feature *Parent) { F.setParent(Parent); }
 
-  static void addChild(Feature &F, Feature &Child) {
+  static void addChild(Feature &F, Feature *Child) {
```
```suggestion
  static void addChild(Feature &F, Feature &Child) {
```
There is no go way to handle nullptrs here, so why should `Child` be a ptr?

s9latimm replied:
As we talked about it last week already -- there is no reason, as any `nullptr` should be handled beforehand -- will revert that and update references accordingly.

vulder commented on lib/Feature/FeatureModel.cpp:
```diff
@@ -14,28 +14,15 @@ void FeatureModel::dump() const {
+Feature *FeatureModel::addFeature(std::unique_ptr<Feature> NewFeature) {
   auto FeatureName = std::string(NewFeature->getName());
   if (!Features.try_emplace(FeatureName, std::move(NewFeature)).second) {
-    return false;
+    return nullptr;
   }
   auto *InsertedFeature = Features[FeatureName].get();
```
If I remember correctly `try_emplace` returns and iterator, bool pair.
You create a local var with this and then use it to make the check and you don't have to lookup the feature.
```
auto PosInsertedFeature = Features.try_emplace(...)
if (PosInsertedFeature.second) {
  return nullptr;
}
...
```

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -30,7 +30,9 @@ bool FeatureModelXmlParser::parseConfigurationOption(xmlNode *Node,
         Opt = Cnt == "True";
       } else if (!xmlStrcmp(Head->name, XmlConstants::PARENT)) {
-        FMB.addParent(Name, Cnt);
+        // TODO(s9latimm) fix me -- breaks if Name not set / parsed yet
```
What should we do in this case?

s9latimm replied:
We could search `name`  beforehand with an initial pass over all `Node->children`, unfortunately that way we have to look at each node twice (or at least until we find `name`). I will add a separate PR from `vara-dev` with the fix, so that we can resolve this quickly.

vulder commented on unittests/Feature/FeatureModel.cpp:
```diff
@@ -48,14 +50,9 @@ TEST(FeatureModel, cloneRoot) {
-  auto CS = Feature::NodeSetType();
-  CS.insert(&AA);
-  CS.insert(&AB);
-  B.addParent("aa", "root")->addFeature(AA);
-  B.addParent("ab", "root")->addFeature(AB);
```
I guess the parent relation is not strictly needed in this test or?

s9latimm replied:
As we add those features without changing root, those edges are already implied (so is the existence of a root node, which it neither of those two).

vulder commented on unittests/Feature/FeatureModelTransaction.cpp:
```diff
@@ -16,42 +16,58 @@ class FeatureModelModificationTest : public ::testing::Test,
 
   // Dummy method to fullfill the FeatureModelModification interface
-  void exec(FeatureModel &FM) override {};
+  void exec(FeatureModel &FM) override{};
```
```suggestion
  void exec(FeatureModel &FM) override {};
```
What version of clang-format are you using?

s9latimm replied:
I updated today to `13.0.0` (latest from `llvm-project/main`) but above change is still persistent (with even more similar diffs according to changed WS rules).

vulder replied:
hmm, kk

vulder reviewed (changes requested):
Did you adapt the FeatureModelBuilder to also use the modifications? Where there problems?

s9latimm committed:
Requested changes

M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModel.cpp
